homeassistant/components/homewizard_energy/config_flow.py

```diff
@@ -0,0 +1,182 @@
+"""Config flow for Homewizard Energy."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import aiohwenergy
+from aiohwenergy.hwenergy import SUPPORTED_DEVICES
+import async_timeout
+from voluptuous import Required, Schema
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_IP_ADDRESS, CONF_NAME, CONF_PORT
+from homeassistant.data_entry_flow import AbortFlow, FlowResult
+
+from .const import CONF_PRODUCT_NAME, CONF_PRODUCT_TYPE, CONF_SERIAL, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for P1 meter."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Set up the instance."""
+        _LOGGER.debug("config_flow __init__")
+
+        self.entry_info = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initiated by the user."""
+
+        _LOGGER.debug("config_flow async_step_user")
+
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=Schema(
+                    {
+                        Required(CONF_IP_ADDRESS): str,
+                    }
+                ),
+                errors=None,
+            )
+
+        device_info = await self._async_try_connect_and_fetch(
+            user_input[CONF_IP_ADDRESS]
+        )
+
+        entry_info = {
+            CONF_IP_ADDRESS: user_input[CONF_IP_ADDRESS],
+            CONF_PORT: 80,
+            CONF_PRODUCT_TYPE: device_info[CONF_PRODUCT_TYPE],
+            CONF_PRODUCT_NAME: device_info[CONF_PRODUCT_NAME],
+            CONF_SERIAL: device_info[CONF_SERIAL],
+        }
+
+        # Add entry
+        return await self._async_create_entry(entry_info)
+
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle zeroconf discovery."""
+
+        _LOGGER.debug("config_flow async_step_zeroconf")
+
+        # Validate doscovery entry
+        if (
+            "host" not in discovery_info
+            or "api_enabled" not in discovery_info["properties"]
+            or "path" not in discovery_info["properties"]
+            or "product_name" not in discovery_info["properties"]
+            or "product_type" not in discovery_info["properties"]
+            or "serial" not in discovery_info["properties"]
```

Comment:
You alraedy have serial here, set it and let it abort right away if it's already configured to avoid having to try to connect.

Reply:
https://github.com/home-assistant/core/pull/55812/commits/c7c7ceb415e7e37ce43bdaf92d6a271fe484f177